lib/execution_engine2/sdk/EE2Runjob.py

```diff
@@ -702,20 +710,95 @@ def _get_run_job_params_from_existing_job(job: Job, user_id: str) -> Dict:
         # Then the next fields are job inputs top level requirements, app run parameters, and scheduler resource requirements
         return run_job_params
 
+    def _check_ws_perms(
+        self, runjob_params, new_batch_job, batch_params, as_admin=False
+    ):
+        """
+        Check a single job, a single batch job, or a retry_multiple request with a mix of different jobs.
+        """
+        if as_admin:
+            return self.sdkmr.check_as_admin(requested_perm=JobPermissions.WRITE)
+        # Batch Param runs
+        if new_batch_job:
+            if batch_params:
+                return self._check_workspace_permissions(batch_params.get("wsid"))
+        # Single job runs
+        elif isinstance(runjob_params, dict):
+            return self._check_workspace_permissions(runjob_params.get("wsid"))
+        # Multiple independent job runs, think retry_multiple()
+        elif isinstance(runjob_params, list):
+            return self._check_workspace_permissions_list(
+                [job_param.get("wsid") for job_param in runjob_params]
+            )
+        else:
+            raise IncorrectParamsException(
+                "runjob_params must be a mapping or a list of mappings"
+            )
+
+    @staticmethod
+    def _propagate_wsid_for_new_batch_jobs(runjob_params, batch_params, new_batch_job):
+        """
+        For batch jobs, check to make sure the job params do not provide a wsid other than None
+        Then Modify the run job params to use the batch params wsid, which may be set to None
+        """
+        if batch_params is None:
+            batch_params = {}
+
+        if new_batch_job:
+            batch_wsid = batch_params.get("wsid")
+            for runjob_param in runjob_params:
+                if runjob_param.get("wsid") is not None:
+                    raise InvalidParameterForBatch()
+                # Do we do a deepcopy here in case the params point to the same obj?
```

Comment:
The problem is that if its the same params object, then next iteration would throw an exception
params = {}
a = params
b = params
_propagate_wsid_for_new_batch_jobs([a,b])
After the first loop, both a and b have 'wsid' = the batch wsid, and the 2nd loop it would raise InvalidParameterForBatch() since B now "specifies" a wsid

Reply:
I see, ok. I assume that can't happen under normal usage though since the input is deserialized from JSON. It'd have to be someone calling against internals with that specific setup, so unlikely, right?

Reply:
It was happening during tests, which caused me to re-write some of the tests. It was already happening. In the future it's likely another test can make this mistake if they are not careful.

Reply:
In which case they'll get a failure, I assume, so that seems ok